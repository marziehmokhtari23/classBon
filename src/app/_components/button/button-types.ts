import { ButtonHTMLAttributes } from "react"
import {ComponentBase} from "@/app/_components/types/component-base-type"
import { LoadingBehavior } from "../types/loading-type"
export  type ButtonProps=ButtonHTMLAttributes<HTMLButtonElement> & ComponentBase & LoadingBehavior & {
isOutline?:boolean,
variant?:string,
animatedIcon?:boolean,
shape?:ButtonShape,
isLink?:boolean,
}
export type ButtonShape="default"|"wide" |"full" | "square"