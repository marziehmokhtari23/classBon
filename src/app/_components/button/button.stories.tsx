import { Meta, StoryObj } from "@storybook/react";
import { Button } from "./button";

const meta: Meta<typeof Button> = {
    component: Button,
    tags: ["autodocs"],
    decorators: [
        (Story) => {
          document.documentElement.classList.add('dark');
          return <Story />;
        },
      ],
}

export default meta;
type Story = StoryObj<typeof Button>;
export const BrandColors: Story = {
    render: () => (
        <div className="flex gap-3">
            <Button>Default</Button>
            <Button variant="neutral">Neutral</Button>
            <Button variant="primary">Primary</Button>
            <Button variant="secondary">Secondary</Button>
            <Button variant="accent">Accent</Button>
            <Button variant="ghost">Ghost</Button>
            
        </div>
    ),
};
export const ButtonSizes:Story={
    render:()=>(
<div className="flex gap-3 items-center">
    <Button size="large">large</Button>
    <Button size="tiny">tiny</Button>
    <Button size="small">small</Button>
    <Button size="normal">normal</Button>
</div>
    )
}
export const OutLineStateButtons:Story={
    render:()=>(
        <div className="flex gap-3 items-center">
<Button isOutline={true} variant="primary">primary</Button>
<Button isOutline={true} variant="warning">primary</Button>
<Button isOutline={true} variant="success">primary</Button>
<Button isOutline={true} variant="error">primary</Button>
<Button isOutline={true} variant="neutral">primary</Button>
        </div>
    
    )
      
}
export const ButtonShapes:Story={
    render:()=>(
        <div className="flex flex-col gap-3">
            <Button shape="full" variant="neutral">full</Button>
            <Button shape="square" variant="neutral">x</Button>
            <Button shape="wide" variant="neutral">wide</Button>
            
        </div>
    )
}
export const DisabledButton:Story={
    render:()=>(
        <Button disabled={true} variant="primary">disabled</Button>
    )
    
    
}




