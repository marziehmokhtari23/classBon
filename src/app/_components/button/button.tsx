import { FC } from "react"
import { ButtonProps } from "./button-types"
import classNames from "classnames"
import { Size } from "../types/size-limit-type"
import { ButtonShape } from "./button-types"
import { Loading } from "../loading"
export const Button: FC<ButtonProps> = ({
    variant,
    isLoading = false,
    isOutline = false,
    animatedIcon = false,
    type = "button",
    children,
    isDisabled = false,
    loadingText = "در حال ارسال درخواست",
    loadingType = "spinner",
    size = "normal",
    shape = "default",
    isLink = false,
    className,
    ...rest
}: ButtonProps) => {
    const sizeClasses: Record<Size, string> = {
        normal: '',
        tiny: 'xs',
        small: 'sm',
        large: 'lg'

    }
    const shapeClasses: Record<ButtonShape, string> = {
        square: 'square',
        default: '',
        wide: 'wide',
        full: 'full',
    }
    const classes = classNames('btn', { 'btn-outline': isOutline }, { 'btn-link': isLink }, { 'animated-icon': animatedIcon }, { "pointer-events-none opacity-80": isLoading }, { [`btn-${variant}`]: variant }, { [`btn-${sizeClasses[size]}`]: size },{ [`btn-${shapeClasses[shape]}`]: shape }, className)
    return (
        <button type={type} disabled={isDisabled}{...rest} className={classes}>
            {isLoading && <Loading type={loadingType}/>}
            {isLoading ? loadingText:children}
        </button>
    )
}