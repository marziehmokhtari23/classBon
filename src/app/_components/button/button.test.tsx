import {describe} from "node:test"

import {render,screen} from '@testing-library/react'
import { Button } from "./button" 
describe('Button compoment',()=>{
    test('default button',()=>{
     const {getByText} = render(<Button>click here</Button>)
     expect(getByText('click here')).toBeInTheDocument();
    })
    test('disabled button',()=>{
        const {getByText} = render(<Button isDisabled={true}>click here</Button>)
        expect(getByText('click here')).toBeDisabled();
       })
    test('variant attribute of button',()=>{
        const {getByText} = render(<Button variant="primary">click here</Button>)
        expect(screen.getByRole('button')).toHaveClass('btn-primary')
       })
       test('isLink props for Button',()=>{
        render(<Button isLink={true}>click here</Button>)
        expect(screen.getByRole('button')).toHaveClass('btn-link')
       })
       test('disabled class for button',()=>{
        render(<Button isLoading={true}>clik here</Button>)
        expect(screen.getByRole('button')).toHaveClass('pointer-events-none opacity-80')
       })
       test('rendering button',()=>{
        render(<Button variant="primary" isLoading={true} isLink={true} isDisabled={true}>click here</Button>)
        screen.debug()
       })
       test('snapShot of button',()=>{
        const {container}=render(<Button variant="primary" isDisabled={true}>click here</Button>)
        expect(container).toMatchSnapshot(`
            <button type={type} disabled={isDisabled}{...rest} className={classes}>
            {isLoading && <Loading type={loadingType}/>}
            {isLoading ? loadingText:children}
        </button>  
            `
        )
       })
})