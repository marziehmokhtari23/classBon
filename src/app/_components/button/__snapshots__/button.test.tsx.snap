// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapShot of button: 
            <button type={type} disabled={isDisabled}{...rest} className={classes}>
            {isLoading && <Loading type={loadingType}/>}
            {isLoading ? loadingText:children}
        </button>  
             1`] = `
<div>
  <button
    class="btn btn-primary btn- btn-"
    disabled=""
    type="button"
  >
    click here
  </button>
</div>
`;
